File: src/test/java/io/outfox/typescriptpoet/test/DecoratorSpecTests.kt
package io.outfox.typescriptpoet.test

import io.outfox.typescriptpoet.CodeBlock
import io.outfox.typescriptpoet.CodeWriter
import io.outfox.typescriptpoet.DecoratorSpec
import io.outfox.typescriptpoet.SymbolSpec
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.CoreMatchers.hasItems
import org.hamcrest.MatcherAssert.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.io.StringWriter


@DisplayName("DecoratorSpec Tests")
class DecoratorSpecTests {

  @Test
  @DisplayName("Generate inline")
  fun testGenInline() {
    val testDec = DecoratorSpec.builder("test")
       .addParameter("value", "100")
       .addParameter("value2", "20")
       .build()

    val out = StringWriter()
    testDec.emit(CodeWriter(out), inline = true)

    assertThat(
       out.toString(),
       equalTo(
          """
            @test(/* value */ 100, /* value2 */ 20)
          """.trimIndent()
       )
    )
  }

  @Test
  @DisplayName("Generate expanded")
  fun testGenExpanded() {
    val testDec = DecoratorSpec.builder("test")
       .addParameter("value", "100")
       .addParameter("value2", "20")
       .build()

    val out = StringWriter()
    testDec.emit(CodeWriter(out), inline = false)

    assertThat(
       out.toString(),
       equalTo(
          """
            @test(
              /* value */ 100,
              /* value2 */ 20
            )
          """.trimIndent()
       )
    )
  }

  @Test
  @DisplayName("Generate with no-argument")
  fun testGenNoArg() {
    val testDec = DecoratorSpec.builder("test")
       .build()

    val out = StringWriter()
    testDec.emit(CodeWriter(out), inline = false)

    assertThat(
       out.toString(),
       equalTo(
          """
            @test
          """.trimIndent()
       )
    )
  }

  @Test
  @DisplayName("Generate factory with no-argument")
  fun testGenNoArgFactory() {
    val testDec = DecoratorSpec.builder("test")
       .asFactory()
       .build()

    val out = StringWriter()
    testDec.emit(CodeWriter(out), inline = false)

    assertThat(
       out.toString(),
       equalTo(
          """
            @test()
          """.trimIndent()
       )
    )
  }

  @Test
  @DisplayName("toBuilder copies all fields")
  fun testToBuilder() {
    val testDecBldr = DecoratorSpec.builder("test")
       .addParameter("value", "100")
       .addParameter("value2", "20")
       .asFactory()
       .build()
       .toBuilder()

    assertThat(testDecBldr.name, equalTo(SymbolSpec.from("test")))
    assertThat(testDecBldr.parameters, hasItems<Pair<String?, CodeBlock>>(Pair("value", CodeBlock.of("100")), Pair("value2", CodeBlock.of("20"))))
    assertThat(testDecBldr.factory, equalTo(true))
  }

}

File: src/__tests__/DecoratorSpec-tests.ts
import { CodeWriter } from "../CodeWriter";
import { DecoratorSpec } from "../DecoratorSpec";
import { StringBuffer } from "../StringBuffer";
import { SymbolSpec } from "../SymbolSpecs";

describe("DecoratorSpec", () => {
  it("generate inline", () => {
    const testDec = DecoratorSpec.create("test")
      .addParameter("value", "100")
      .addParameter("value2", "20");
    expect(emit(testDec, true)).toMatchInlineSnapshot(
      `"@test(/* value */ 100, /* value2 */ 20)"`
    );
  });

  it("generate expanded", () => {
    const testDec = DecoratorSpec.create("test")
      .addParameter("value", "100")
      .addParameter("value2", "20");
    expect(emit(testDec)).toMatchInlineSnapshot(`
"@test(
  /* value */ 100,
  /* value2 */ 20
)"
`);
  });

  it("generate with no-argument", () => {
    const testDec = DecoratorSpec.create("test");
    expect(emit(testDec)).toMatchInlineSnapshot(`"@test"`);
  });

  it("generate factory with no-argument", () => {
    const testDec = DecoratorSpec.create("test").asFactory();
    expect(emit(testDec)).toMatchInlineSnapshot(`"@test()"`);
  });
});

function emit(spec: DecoratorSpec, inline: boolean = false): string {
  const out = new StringBuffer();
  spec.emit(new CodeWriter(out), inline);
  return out.toString();
}
